src/uct/ib/rc/accel/rc_mlx5_common.h

```diff
@@ -406,81 +404,78 @@ uct_rc_mlx5_txqp_dptr_post(uct_rc_iface_t *iface, enum ibv_qp_type qp_type,
         ucs_assert(length < (2ul << 30));
         ucs_assert(length <= iface->super.config.seg_size);
 
-        wqe_size         = ctrl_av_size + sizeof(*dptr);
-        dptr             = uct_ib_mlx5_get_next_seg(txwq, ctrl, ctrl_av_size);
-        uct_ib_mlx5_set_data_seg(dptr, buffer, length, *lkey_p);
+        wqe_size = ctrl_av_size + sizeof(struct mlx5_wqe_data_seg);
+        uct_ib_mlx5_set_data_seg(next_seg, buffer, length, *lkey_p);
         break;
 
     case MLX5_OPCODE_RDMA_READ:
     case MLX5_OPCODE_RDMA_WRITE:
         /* Set RDMA segment */
         ucs_assert(length <= UCT_IB_MAX_MESSAGE_SIZE);
 
-        raddr            = uct_ib_mlx5_get_next_seg(txwq, ctrl, ctrl_av_size);
+        raddr = next_seg;
         uct_ib_mlx5_ep_set_rdma_seg(raddr, remote_addr, rkey);
 
         /* Data segment */
         if (length == 0) {
             wqe_size     = ctrl_av_size + sizeof(*raddr);
         } else {
+            /* dptr cannot wrap, because ctrl+av could be either 2 or 4 segs */
+            dptr         = uct_ib_mlx5_txwq_wrap_none(txwq, raddr + 1);
             wqe_size     = ctrl_av_size + sizeof(*raddr) + sizeof(*dptr);
-            uct_ib_mlx5_set_data_seg((void*)(raddr + 1), buffer, length, *lkey_p);
+            uct_ib_mlx5_set_data_seg(dptr, buffer, length, *lkey_p);
         }
         break;
 
     case MLX5_OPCODE_ATOMIC_FA:
     case MLX5_OPCODE_ATOMIC_CS:
         ucs_assert(length == sizeof(uint64_t));
-        raddr = uct_ib_mlx5_get_next_seg(txwq, ctrl, ctrl_av_size);
+        raddr = next_seg;
```

Comment:
It might be better to replace `next_seg` by `raddr` and delete `next_seq` variable.